Userland/Libraries/LibGfx/QOIWriter.cpp

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2022, Olivier De Cannière <olivier.decanniere96@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "QOIWriter.h"
+#include <AK/String.h>
+
+namespace Gfx {
+
+ByteBuffer QOIWriter::encode(Bitmap const& bitmap)
+{
+    QOIWriter writer;
+    writer.add_header(bitmap.width(), bitmap.height(), Channels::RGBA, Colorspace::sRGB);
+
+    Color previous_pixel = { 0, 0, 0, 255 };
+
+    bool creating_run = false;
+    int run_length = 0;
+
+    for (int y = 0; y < bitmap.height(); y++) {
+        for (int x = 0; x < bitmap.width(); x++) {
+            Color pixel = bitmap.get_pixel(x, y);
+
+            // Check for at most 62 consecutive identical pixels
+            if (pixel.red() == previous_pixel.red() && pixel.green() == previous_pixel.green() && pixel.blue() == previous_pixel.blue() && pixel.alpha() == previous_pixel.alpha()) {
+                if (!creating_run) {
+                    creating_run = true;
+                    run_length = 0;
+                    writer.insert_into_running_array(pixel);
+                }
+
+                run_length++;
+
+                // If the run reaches maximum length of 62 or if this is the last pixel create the chunk
+                if (run_length == 62 || (y == bitmap.height() - 1 && x == bitmap.width() - 1)) {
+                    writer.add_run_chunk(run_length);
+                    creating_run = false;
+                }
+
+                continue;
+            }
+
+            // Run ended with the previous pixel. Create a chunk for it and continue processing this pixel
+            if (creating_run) {
+                writer.add_run_chunk(run_length);
+                creating_run = false;
+            }
+
+            // Check if the pixel matches a pixel in the running array
+            auto index = QOIWriter::pixel_hash_function(pixel.red(), pixel.green(), pixel.blue(), pixel.alpha());
```

Comment:
Can this just take a `Color` directly instead of individual values? Also don't need to specify `QOIWriter` here.